### ISSUE TITLE
Rprop - resilient propagation

### ISSUE THREAD

**Issue by Luk0r** [JuliaOpt/Optim.jl#124]
As an enhancement I suggest to implement the Rprop algorithm.
It is a rather simple optimization algorithm only requiring the function value and the gradient.
It is just adjusting the step size dependent on the gradient sign change ( increase step size if gradient sign is not changing, decrease if gradient sign changed; for each dimension separately)

See this paper for a detailed description (with simple pseudo code):
http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.17.1332

And see this implementation in C++ (Rprop.cpp, Irpropplus.cpp, ...):
https://github.com/Luk0r/gaupro/tree/master/optimizer/src

I used it for optimizing the hyperparameters of a high dimensional Gaussian Process.

**Comment by johnmyleswhite** [comment]
Care to submit a pull request?

**Comment by Luk0r** [comment]
I just started with julia and played with Gaussian Process implementation, that brought me to enhancement idea.
I'm completely not familiar with julia ;>

**Comment by johnmyleswhite** [comment]
Ok. I think I may close this issue then as I think Rprop doesn't seem well-suited to Optim's current design since Rprop seems to be focused on stochastic optimization, whereas Optim is only meant to optimize deterministic functions.

**Comment by Luk0r** [comment]
I see what you mean. But imho don't limit the usefulness of the package by such constraints. Optimization is such a wide and open field and so many algorithms/people/problems need it in one or another way.
If it fits the interface I would take as many algorithms as possible into account.

also found a nice vid visualizing gradient descent and Rprop:
https://www.youtube.com/watch?v=Cy2g9_hR-5Y

**Comment by johnmyleswhite** [comment]
Unfortunately, it doesn't fit the interface: there's no way to express stochastic optimization efficiently using this framework. You need custom code for that. See something like StreamsStats for an example.
